### DISCUSSION TITLE
How does Successive Halving work?

### DISCUSSION THREAD

**Discussion by IlmarsKuht** [optuna/optuna#4236] [Q&A]
So from my understating of successive halving is that X amount of trials are completed up to a certain point, for example 10 epochs.
Then the top half of those trials are chosen and then the X/2 trials are trained for 20 epochs. Correct me if I am wrong.

But this is not how optuna does it, optuna goes through the whole first trial and second trial before pruning each trial one by one.

So I am confused where is the tournament like system? there is only one round.

Note: I am a beginner at this, please bash me until it goes into my head.

**Comment by Alnusjaponica** [comment]
Your understanding of the pure [Successive Halving](http://proceedings.mlr.press/v28/karnin13.pdf) seems to be correct. However, what is implemented as [optuna.pruners.SuccessiveHalvingPruner](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.SuccessiveHalvingPruner.html#optuna.pruners.SuccessiveHalvingPruner) is a little different from that. 
The class is mainly based on [Asynchronous Successive Halving](https://proceedings.mlsys.org/paper/2020/hash/f4b9ec30ad9f68f89b29639786cb62ef-Abstract.html), which supports distributed computing settings. The algorithm is also modified for Optuna design, since it does not support suspending or resuming ongoing trials.
